### ISSUE TITLE
Fairseq Multi-GPU support

### ISSUE THREAD

**Issue by stephenroller** [facebookresearch/ParlAI#906]
(Depends on #876 landing first.)

Right now FairseqAgent hardcodes to only use one GPU, which gives up some of the best aspects of FairSeq, and @uralik potentially needs this supported. 

https://github.com/facebookresearch/ParlAI/blob/68552610b7068b88d9c92f615cdf83104a77998b/parlai/agents/fairseq/fairseq.py#L78-L81

Approach:
- [ ] Drop those hardcoded lines
- [ ] Add a call to `fairseq.options.add_distributed_training_args` in `add_cmdline_args`
- [ ] ???
- [ ] Profit

There's some trouble (hard crash) if this is not hard coded, but the exact nature of this crash isn't clear at the moment -- it looks like fairseq expects multigpu support to be performed using multiple processes:

```python-traceback
Traceback (most recent call last):
  File "examples/train_model.py", line 28, in <module>
    TrainLoop(opt).train()
  File "/private/home/roller/working/parlai/parlai/scripts/train_model.py", line 272, in train
    world.parley()
  File "/private/home/roller/working/parlai/parlai/core/worlds.py", line 628, in parley
    batch_act = self.batch_act(agent_idx, batch_observations[agent_idx])
  File "/private/home/roller/working/parlai/parlai/core/worlds.py", line 601, in batch_act
    batch_actions = a.batch_act(batch_observation)
  File "/private/home/roller/working/parlai/parlai/agents/fairseq/fairseq.py", line 308, in batch_act
    self.trainer.train_step(samples)
  File "/private/home/roller/working/forks/fairseq/fairseq/trainer.py", line 137, in train_step
    (sample_sizes, logging_outputs, ooms_fwd, ooms_bwd)
  File "/private/home/roller/working/forks/fairseq/fairseq/distributed_utils.py", line 57, in all_gather_list
    world_size = torch.distributed.get_world_size()
  File "/private/home/roller/.conda/envs/chat/lib/python3.6/site-packages/torch/distributed/__init__.py", line 169, in get_world_size
    assert torch.distributed._initialized
AssertionError
```

This looks like it can be handled by `fairseq.distributed_utils.distributed_init`, but the general process desires something that looks like `fairseq.multiprocessing_train` which spawns multiple processes. It's not clear to me how this will work with ParlAI right now.

**Comment by stephenroller** [comment]
Looked into doing this using DataParallel. It requires an upstream change to fairseq. The semantics of all this are likely to change in a future release.

**Comment by stephenroller** [comment]
This is implemented in spirit with #1319, although that doesn't *specifically* help Fairseq at the moment.
